python/lsst/pipe/base/graph/_loadHelpers.py

```diff
@@ -107,58 +108,87 @@ class DefaultLoadHelper:
+        fmtSize = struct.calcsize(STRUCT_FMT_BASE)
         preambleSize = magicSize + fmtSize
 
         headerBytes = self._readBytes(0, preambleSize)
         magic = headerBytes[:magicSize]
-        sizeBytes = headerBytes[magicSize:]
+        versionBytes = headerBytes[magicSize:]
 
         if magic != MAGIC_BYTES:
             raise ValueError("This file does not appear to be a quantum graph save got magic bytes "
                              f"{magic}, expected {MAGIC_BYTES}")
 
         # Turn they encode bytes back into a python int object
-        save_version, taskDefSize, nodeSize = struct.unpack('>HQQ', sizeBytes)
+        save_version, = struct.unpack(STRUCT_FMT_BASE, versionBytes)
+
+        if save_version > SAVE_VERSION:
+            raise RuntimeError(f"The version of this save file is {save_version}, but this version of"
+                               f"Quantum Graph software only knows how to read up to version {SAVE_VERSION}")
+
+        # read in the next bits
+        fmtString = STRUCT_FMT_STRING[save_version]
+        infoSize = struct.calcsize(fmtString)
+        infoBytes = self._readBytes(preambleSize, preambleSize+infoSize)
+        infoUnpack = struct.unpack(fmtString, infoBytes)
+
+        preambleSize += infoSize
 
         # Store the save version, so future read codes can make use of any
         # format changes to the save protocol
         self.save_version = save_version
 
-        return preambleSize, taskDefSize, nodeSize
+        return preambleSize, infoUnpack
 
-    def _readByteMappings(self, preambleSize, headerSize, taskDefSize):
+    def _readByteMappings(self, preambleSize, headerSize, infoSize):
         # Take the header size explicitly so subclasses can modify before
         # This task is called
 
         # read the bytes of taskDef bytes and nodes skipping the size bytes
         headerMaps = self._readBytes(preambleSize, headerSize)
 
-        # read the map of taskDef bytes back in skipping the size bytes
-        self.taskDefMap = pickle.loads(headerMaps[:taskDefSize])
+        if self.save_version == 1:
+            taskDefSize, _ = infoSize
+
+            # read the map of taskDef bytes back in skipping the size bytes
+            self.taskDefMap = pickle.loads(headerMaps[:taskDefSize])
+
+            # read back in the graph id
+            self._buildId = self.taskDefMap['__GraphBuildID']
 
-        # read back in the graph id
-        self._buildId = self.taskDefMap['__GraphBuildID']
+            # read the map of the node objects back in skipping bytes
+            # corresponding to the taskDef byte map
+            self.map = pickle.loads(headerMaps[taskDefSize:])
 
-        # read the map of the node objects back in skipping bytes
-        # corresponding to the taskDef byte map
-        self.map = pickle.loads(headerMaps[taskDefSize:])
+            # There is no metadata for old versions
+            self.metadata = None
+        elif self.save_version >= 2:
```

Comment:
I would probably be more defensive here and only check for `self.save_version == 2` and add `else: raise ...` to force explicit verification when new version is added. Same for similar check in `__init__`.

Reply:
I went back and forth on this. I ended up settling on this pattern for 2 reasons.
1) People would be editing the branch anyway
2) I could not predict where someone would want to branch the logic, i.e.
```
elif self.save_version >= 2:
    <some shared logic>
    if self.save_version == 3:
        <do some logic extra for version 3>
```
or
```
elif self.save_version == 2:
    <some version 2 logic>
elif self.save_version >= 3:
    <some new logic all together>
```
I normally would not assume that we should force someone to review various parts of code that might not be related to their change, but in this case I can see your point.